DMCompiler/DM/Expressions/Dereference.cs

```diff
@@ -114,6 +114,13 @@ public Dereference(Location location, DreamPath? path, DMExpression expression,
                     break;
 
                 case IndexOperation indexOperation:
+                    if (NestedPath is not null) {
+                        var obj = DMObjectTree.GetDMObject(NestedPath.Value, false);
+                        if (obj is not null && obj.IsSubtypeOf(DreamPath.Datum) && !obj.HasProc("operator[]")) {
+                            DMCompiler.Emit(WarningCode.InvalidIndexOperation, Location, "Invalid index operation. datum[] index operations are not valid starting in BYOND 515.1641");
+                        }
+                    }
```

Comment:
This should also be done in `EmitReference()` so that we can emit a warning for `datum["foo"] = value`.

Reply:
don't forget to check for `operator[]=` instead